api/geyser/src/main/java/org/geysermc/geyser/api/custom/items/CustomItemData.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 2019-2022 GeyserMC. http://geysermc.org
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @author GeyserMC
+ * @link https://github.com/GeyserMC/Geyser
+ */
+
+package org.geysermc.geyser.api.custom.items;
+
+import org.geysermc.geyser.api.custom.CustomRenderOffsets;
+
+/**
+ * This is used to store data for a custom item.
+ */
+public class CustomItemData {
+    private CustomItemRegistrationType registrationType;
+
+    private Integer customModelData;
```

Comment:
Can/should an item have multiple predicates?

Reply:
Well no, that’s why there’s a registration type. And you can’t set two of them.

Reply:
Then I don't think we should have a boxed Integer and Double fields here (one will always be null). Can we instead have a class that determines when the custom item data should be registered? Something like CustomItemRegistrationType, but with overload classes that specify how.